File: jsonrpc.py

import json
from websocket_server import WebsocketServer

class Handler:
    def _onopen(session):
        pass

    def _onclose(session):
        pass

class RpcSession:
    def __init__(self, wsock, handler):
        self.socket   = wsock        # 与此RpcSession关联的WebSocketSession
        self.handler  = handler      # RPC处理器
        self.callback = {}           # callback与id的关联字典
        self.id       = 1            # id计数器
        wsock._rpcses  = self        # WebSocketSession中与自身相关联的RpcSession

    # 处理call
    def _call(self, id, name, data):
        try:
            fn = getattr(self.handler, name)
            data = fn(self, data)
            data = data if data else []
        except AttributeError:
            data = []
        if id:
            self.retn(id, data)

    # 处理return
    def _retn(self, id, data):
        callback = self._callback4id(id)
        if callback:
            callback(self, data)

    def _handle(self, msg):
        # 解析收到的数据
        msg = json.loads(msg)
        if type(msg[0]) is str:     # call
            name = msg[0]
            id   = msg[1]
            self._call(id, name, msg[2])
        else:                       # return
            id = msg[0]
            self._retn(id, msg[1])

    # 获取callback关联的id
    def _id4callback(self, callback):
        if not callback:
            return 0
        self.callback[self.id] = callback
        id = self.id
        self.id += 1
        return id

    # 获取id关联的callback
    def _callback4id(self, id):
        if id in self.callback:
            callback = self.callback[id]
            del self.callback[id]
            return callback

    # call
    def call(self, name, args = [], callback = None):
        assert type(name) is str
        id = self._id4callback(callback)
        msg = [name, id, args]
        self.socket.send(json.dumps(msg))

    # return
    def retn(self, id, args):
        msg = [id, args]
        self.socket.send(json.dumps(msg))

class RpcServer(WebsocketServer):
    def __init__(self, handler = Handler, **args):
        WebsocketServer.__init__(self, **args)
        self.handler = handler
        self.sessions = {}

    def onopen(self, socket):
        rpcsession = RpcSession(socket, self.handler)
        self.sessions[rpcsession] = 1
        try:
            self.handler._onopen(rpcsession)
        except Exception:
            pass

    def onmessage(self, session, msg):
        session._rpcses._handle(msg)

    def onclose(self, socket):
        rpcsession = socket._rpcses
        try:
            self.handler._onclose(socket._rpcses)
            del self.sessions[rpcsession]
        except Exception:
            pass

    def call(self, name, args):
        for ss in self.sessions.keys():
            ss.call(name, args)

    def start(self):
        self.run_forever()

File: web/jsonrpc.js

function Session (url, handler) {
    var socket = new WebSocket(url);
    var self = this;
    socket.onopen = function (event) {
        handler._onopen(self);
    }
    socket.onmessage = function (event) {
        var data = JSON.parse(event.data);
        self._handle(data);
    }
    socket.onerror = function (event) {
        handler._onerror(self);
    }
    socket.onclose = function (event) {
        handler._onclose(self);
    }
    socket.session = this;
    this.socket = socket;           // Session关联的WebSocket
    this.handler = handler;         // Session的处理器
    this.id = 1;                    // id计数器
    this.callback = {};             // id callback 关联表
    return this;
}
Session.prototype.close = function () {
    this.socket.close();
}
// 通过callback获取id
Session.prototype._id4callback = function (callback) {
    if (callback) {
        this.callback[this.id] = callback;
        return this.id++;
    }
    return null;
}
// 通过id获取callback
Session.prototype._callback4id = function (id) {
    var callback = this.callback[id];
    if (callback) {
        delete this.callback[id];
        return callback;
    }
    return null;
}
// 处理数据包
Session.prototype._handle = function (data) {
    if (typeof data[0] == 'string') {
        var name = data[0];
        var id = data[1];
        this._call(id, name, data[2]);
    } else {
        var id = data[0];
        this._retn(id, data[1]);
    }
}
// 处理call
Session.prototype._call = function (id, name, data) {
    var fn = this.handler[name];
    var ret = null;
    if (fn) {
        ret = fn(self, data);
        if (!data) data = null;
    }
    if (id)
        this.retn(id, ret);
}
// 处理return
Session.prototype._retn = function (id, data) {
    var callback = this._callback4id(id);
    if (callback)
        callback(this, data);
}
// call
Session.prototype.call = function (name, args, callback) {
    var id = this._id4callback(callback);
    var msg = [name, id, args];
    this.socket.send(JSON.stringify(msg));
}
// return
Session.prototype.retn = function (id, args) {
    var msg = [id, args];
    this.socket.send(JSON.stringify(msg));
}
